zerver/lib/send_email.py

```diff
@@ -416,23 +417,20 @@ def clear_scheduled_invitation_emails(email: str) -> None:
 
 
 @transaction.atomic(savepoint=False)
-def clear_scheduled_emails(user_ids: List[int], email_type: Optional[int] = None) -> None:
+def clear_scheduled_emails(user_id: int, email_type: Optional[int] = None) -> None:
     # We need to obtain a FOR UPDATE lock on the selected rows to keep a concurrent
     # execution of this function (or something else) from deleting them before we access
     # the .users attribute.
-    items = ScheduledEmail.objects.filter(users__in=user_ids).select_for_update()
+    items = (
+        ScheduledEmail.objects.filter(users__in=[user_id])
+        .prefetch_related("users")
+        .select_for_update()
```

Comment:
The change from `users__in=user_ids` to `users__in=[user_id]` is in the wrong commit.

(Does `users__id=user_id` work?)

Have you verified that `select_for_update` also takes locks on the `prefetch_related` tables?

Reply:
Thanks for catching the `users__in=[user_id]` placement.  Fixed and repushed.

> (Does users__id=user_id work?)

Sadly, no.

> Have you verified that select_for_update also takes locks on the prefetch_related tables?

Yes.  I talked a bit about this in the commit message for `scheduled_email: Consistently lock users table`, but I can expound.

`ScheduledEmail.objects.filter(users__in=[1]).prefetch_related("users").select_for_update()` runs the following queries:
```sql
SELECT "zerver_scheduledemail"."id", "zerver_scheduledemail"."scheduled_timestamp", "zerver_scheduledemail"."data", "zerver_scheduledemail"."realm_id", "zerver_scheduledemail"."address", "zerver_scheduledemail"."type" FROM "zerver_scheduledemail" INNER JOIN "zerver_scheduledemail_users" ON ("zerver_scheduledemail"."id" = "zerver_scheduledemail_users"."scheduledemail_id") WHERE "zerver_scheduledemail_users"."userprofile_id" IN (1) FOR UPDATE

SELECT ("zerver_scheduledemail_users"."scheduledemail_id") AS "_prefetch_related_val_scheduledemail_id", "zerver_userprofile"."id", [...] FROM "zerver_userprofile" INNER JOIN "zerver_scheduledemail_users" ON ("zerver_userprofile"."id" = "zerver_scheduledemail_users"."userprofile_id") WHERE "zerver_scheduledemail_users"."scheduledemail_id" IN (5)

SELECT "zerver_realm"."id", [...] FROM "zerver_realm" WHERE "zerver_realm"."id" = 1 LIMIT 21
```

The first `SELECT ... FOR UPDATE` covers `zerver_scheduledemail` and `zerver_scheduledemail_users` tables, but not the `zerver_userprofile`, which is what we want.  Being in the list of tables and joins means it's included in the `FOR UPDATE` lock; running `select * from zerver_scheduledemail_users for update` in a separate transaction while the first SQL statement above has run in an open transaction correctly blocks.